test/api.jl

```diff
@@ -115,5 +116,29 @@ recommendation = provide_recommendation(trial_vector)
 recommendation_multi = provide_recommendation(trial_vector_multihyperparams)
 @test recommendation_multi == Dict(:x => 3, :y => 7)
 
+# Test case reproducing the issue in “obs_memo” function from the higher level
+# More details on this issue see https://github.com/IQVIA-ML/TreeParzen.jl/issues/86
+space = Dict(
+    :a => HP.Choice(:a, [8, 9, 10]),
+    :b => HP.Choice(:b, [1, 2, 3])
+    )
+
+Random.seed!(10)
+trials = TreeParzen.Trials.Trial[]
+
+for i in 1:30
+    trial = ask(space, trials, TreeParzen.Config())
+    tell!(trials, trial, trial.hyperparams[:b]/trial.hyperparams[:a])
+end
+# With this example space and loss function, expected vals for the best param should be different
+# but with the issue in “obs_memo” function, they are equal due to same nid
+@test_broken trials[30].vals[:a] != trials[30].vals[:b]
+
+# To have the smaller loss, the expected best hyperparams[:a] is 10 and hyperparams[:b] is 1
+expected_a = 10
+expected_b = 1
+
+@test_broken (trials[30].hyperparams[:a] == expected_a) && (trials[30].hyperparams[:b] == expected_b)
+
```

Comment:
```suggestion
a_dist_values = [8, 9, 10]
b_dist_values = [1, 2, 3]

space = Dict(
    :a => HP.Choice(:a, a_dist_values),
    :b => HP.Choice(:b, b_dist_values)
    )

trials = TreeParzen.Trials.Trial[]

for i in 1:30
    trial = ask(space, trials, TreeParzen.Config())
    tell!(trials, trial, trial.hyperparams[:b]/trial.hyperparams[:a])
end

# With this example space we expect all samples for :a to come from [8, 9, 10]
# and all samples for :b to come from [1, 2, ,3]
# but with the issue in “obs_memo” function, both will sample
# from the same distribution, which is wrong

samples_a = getindex.(getfield.(trials, Ref(:vals)), Ref(:a))
samples_b = getindex.(getfield.(trials, Ref(:vals)), Ref(:b))

@test all(in.(samples_a, Ref(a_dist_values)))
@test all(in.(samples_b, Ref(b_dist_values)))

```

This is the code which most accurately reflects the properties we want to test -- that samples for A come from A distribution and samples for B come from B distribution. Currently, `samples_a` and `samples_b` will all contain values from `[1, 2, 3]`, which is the bug.

Reply:
NO this wrong 
# With this example space we expect all samples for :a to come from [8, 9, 10]
# and all samples for :b to come from [1, 2, ,3]
# but with the issue in “obs_memo” function, both will sample
# from the same distribution, which is wrong

Reply:
Lets jump on a call if still is not clear

Reply:
<img width="468" alt="image" src="https://user-images.githubusercontent.com/121951394/212744447-325f0f5b-cd13-47b0-b67f-020cd40aed9d.png">
This would help to clarify the reason.

Reply:
> <img alt="image" width="468" src="https://user-images.githubusercontent.com/121951394/212744447-325f0f5b-cd13-47b0-b67f-020cd40aed9d.png">
> This would help to clarify the reason.

This is exactly why the test is right, thanks for demonstrating it.

The values for samples A **should** be in the set [8, 9, 10] (hence `@test_broken`, and when you fix the issue, this will start passing). I'm not really sure what either of you find contentious about this.

Reply:
When the bug is fixed, samples_a and samples_b are also the numbers of [1,2,3]. I think a meeting is required to clarify it.

Reply:
> When the bug is fixed, samples_a and samples_b are also the numbers of [1,2,3]. I think a meeting is required to clarify it.

Is this statement based on the fact that you are referring to the indices 1, 2, 3 that you expect to be returned (and tested) and not the values at those indices? 
Given the prior conversations I understand that the low level bug at `obs_memo` for the 3 given examples [mentioned_here](https://github.com/IQVIA-ML/TreeParzen.jl/pull/87#discussion_r1069594941) is because `obs_memo` returns the same `nid` (vals[nid]) not the parameter[nid].
From the following conversations, on higher level this results in draws being taken from the wrong sets [as_shown_here[(https://github.com/IQVIA-ML/TreeParzen.jl/pull/87#discussion_r1071435630). 
The example test written by @yaxxie looks reasonable and shows there is an issue. 
But I understand that @FatemehTahavori and @BojieSheng are in disagreement that yaxxie's test does not reflect the bug observed. 
In my understanding samples_a and samples_b should be drawn from different sets given that particular test scenario and fixing the lower level bug would/should eventually be reflected on this higher level. This is with the assumption we are testing the parameter values not their indices. If we want to see if the indices returned are as expected, then it's a different test for me (and may be something that can be done on lower level as a sanity check).

Reply:
@yaxxie this test `@test all(in.(samples_b, Ref(b_dist_values)))
` is passing because samples_b is indexes from 1,2 and 3 and the values in space b_dist_values is 1, 2, 3. If you change b_dist_values = [11, 12, 13] this test will break too. The random seed is needed due to posterior function

Reply:
Yes, changing the `b_dist_values` to other values is less misleading given that it is clear now to me we are looking at indices 1, 2, 3 returned and not the values at those indices.  So I can see now why the altered test seems faulty and doesn't compare like for like because it is comparing samples which are indices 1, 2, 3 defined as this `getindex.(getfield.(trials, Ref(:vals)), Ref(:b))` (or `Ref(:a)`) with the distribution values (which are not indices). Therefore, the test `@test all(in.(samples_b, Ref(b_dist_values)))` only happens to past because by coincidence it was chosen for `b_dist_values` to be [1, 2, 3] any other values will not pass.
This is clear now.
So what we need to observe (by failing tests until fixed) that is something similar to what you tried to achieve here `trials[30].vals[:a] != trials[30].vals[:b]` where after the random trials of 20 are done, the remainder indices for `:a` are not equal to `:b` which is currently happening given the bug for 3 identified cases.

Reply:
Ok I see the mistake I made. I'm sorry about that -- I assumed incorrectly that `vals` and `hyperparams` were the same.

However, you still don't need the seed. The `vals` allows you to inspect the properties of the posterior indirectly. When the bug is in effect the last samples for vals :a and vals :b are the same (because sampling the same posteriors)

A more relevant test would be to run 100 trials and say that the last 50 samples have (mostly, say up to a tolerance of 10% the same) different values (we expect the posterior to tend towards indices with lower values for B and higher values for A with the loss function provided). What do you think about this?

I think it is important not to write tests relying on a specific random seed for a package like this because it can obscure the presence of bugs (and we're trying to fix one).

Reply:
I'll resolve this comment since the suggestion is no longer relevant, but I'll ask that you consider implementing a property based test instead of the current test.

Reply:
and then once the correct indices are returned to make sure we get the right values at those indices as well